src/stdatamodels/stnode.py

```diff
@@ -0,0 +1,565 @@
+        update = False
+        errmsg = _error_message(path, error)
+        if pass_invalid_values:
+            update = True
+        if strict_validation:
+            raise jsonschema.ValidationError(errmsg)
+        else:
+            warnings.warn(errmsg, ValidationWarning)
+    return update
+
+def _check_value(value, schema, validator_context):
+    """
+    Perform the actual validation.
+    """
+
+    validator_resolver = validator_context.resolver
+
+    temp_schema = {
+        '$schema':
+        'http://stsci.edu/schemas/asdf-schema/0.1.0/asdf-schema'}
+    print('schema =', schema)
+    temp_schema.update(schema)
+    validator = asdfschema.get_validator(temp_schema,
+                                          validator_context,
+                                          validator_callbacks,
+                                          validator_resolver)
+
+    #value = yamlutil.custom_tree_to_tagged_tree(value, validator_context)
+    validator.validate(value, _schema=temp_schema)
+    validator_context.close()
+
+def _validate(attr, instance, schema, ctx):
+    tagged_tree = yamlutil.custom_tree_to_tagged_tree(instance, ctx)
+    return _value_change(attr, tagged_tree, schema, False, strict_validation, ctx)
+
+def _get_schema_for_property(schema, attr):
+    subschema = schema.get('properties', {}).get(attr, None)
+    print('XXXX _get_schema_for_property', schema)
+    if subschema is not None:
+        return subschema
+    for combiner in ['allOf', 'anyOf']:
+        for subschema in schema.get(combiner, []):
+            subsubschema = _get_schema_for_property(subschema, attr)
+            if subsubschema != {}:
+                return subsubschema
+    return {}
+
+
+class DNode(UserDict):
+
+    _tag = None
+    _ctx = None
+
+    def __init__(self, node=None, parent=None, name=None):
+
+        if node is None:
+            self.__dict__['_data']= {}
+        elif isinstance(node, dict):
+            self.__dict__['_data'] = node
+        else:
+            raise ValueError("Initializer only accepts dicts")
+        self._x_schema = None
+        self._schema_uri = None
+        self._parent = parent
+        self._name = name
+        # else:
+        #     self.data = node.data
+
+    @property
+    def ctx(self):
+        if self._ctx is None:
+            DNode._ctx = asdf.AsdfFile()
+        return self._ctx
+    
+   
+    def __getattr__(self, key):
+        """
+        Permit accessing dict keys as attributes, assuming they are legal Python
+        variable names.
+        """
+        if key in self._data:
+            print('---------- (getattr) key = ', key)
+            value = self._data[key]
+            if isinstance(value, dict):
+                return DNode(value, parent=self, name=key)
+            elif isinstance(value, list):
+                return LNode(value)
+            else:
+                return value
+        else:
+            raise KeyError(f"No such key ({key}) found in node")
```

Comment:
So if a node key like jwst.meta.exposure is missing entirely, and a user attempts to access jwst.meta.exposure.type, they'll receive a `KeyError` and not just a `None` value right?

Reply:
Yes, I believe so. That is the behavior we want now isn't it? Though we should work out how we deal with permitting null as values we handle. In other words, if we want `None`, we should be explicit about it.